src/gfx.cpp

```diff
@@ -154,6 +154,121 @@ void GfxFillRect(int left, int top, int right, int bottom, int colour, FillRectM
+ * @param mode
+ *         FILLRECT_OPAQUE:   Fill the polygon with the specified colour
+ *         FILLRECT_CHECKER:  Like FILLRECT_OPAQUE, but only draw every second pixel (used to grey out things)
+ *         FILLRECT_RECOLOUR:  Apply a recolour sprite to every pixel in the rectangle currently on screen
+ */
+void GfxFillPolygon(const std::vector<Point> &shape, int colour, FillRectMode mode)
+{
+	typedef std::pair<Point, Point> Segment;
+	std::vector<Segment> segments;
+	Blitter *blitter = BlitterFactory::GetCurrentBlitter();
+	const DrawPixelInfo *dpi = _cur_dpi;
+
+	if (dpi->zoom != ZOOM_LVL_NORMAL) return;
+	assert(shape.size() >= 3);
+
+	{
+		Point prev = shape[0];
+		prev.x -= dpi->left;
+		prev.y -= dpi->top;
+		for (size_t sp = 1; sp < shape.size(); ++sp) {
+			Point pt = shape[sp];
+			pt.x -= dpi->left;
+			pt.y -= dpi->top;
+			/* Create segments for all non-horizontal lines in the polygon.
+			 * The segments always have lowest Y coordinate first. */
+			if (prev.y > pt.y) {
+				segments.emplace_back(pt, prev);
+			} else if (prev.y < pt.y) {
+				segments.emplace_back(prev, pt);
+			}
+			prev = pt;
+		}
+		/* Create the final segment connecting last and first point. */
+		Point first = shape[0];
+		first.x -= dpi->left;
+		first.y -= dpi->top;
+		if (prev.y > first.y) {
+			segments.emplace_back(first, prev);
+		} else if (prev.y < first.y) {
+			segments.emplace_back(prev, first);
+		}
+	}
+
+	/* Check that this wasn't an empty shape (all points on a horizontal line.) */
+	if (segments.empty()) return;
+
+	/* Sort the segments by first point Y coordinate. */
+	std::sort(segments.begin(), segments.end(), [](const Segment &a, const Segment &b) { return a.first.y < b.first.y; });
+
+	/* Scan through the segments and paint each scanline. */
+	std::vector<Segment> active;
+	int y = segments.front().first.y;
+	std::vector<Segment>::iterator nextseg = segments.begin();
+	while (!active.empty() || nextseg != segments.end()) {
+		/* Activate all segments starting on this scanline. */
+		while (nextseg != segments.end() && nextseg->first.y == y) {
+			active.push_back(*nextseg);
+			++nextseg;
+		}
+		/* Clean up segments that have ended. */
+		for (size_t a = 0; a < active.size();) {
+			if (active[a].second.y == y) {
+				active.erase(active.begin() + a);
+			} else {
+				++a;
+			}
+		}
+		/* Skip scanlines before clipping rectangle. */
+		if (y < 0) {
+			++y;
+			continue;
+		}
+		/*  Intersect scanline with all active segments. */
+		int x1 = INT_MAX, x2 = 0;
+		for (const Segment &s : active) {
+			const int sdx = s.second.x - s.first.x;
+			const int sdy = s.second.y - s.first.y;
+			const int ldy = y - s.first.y;
+			const int x = s.first.x + sdx * ldy / sdy;
+			x1 = min(x1, x);
+			x2 = max(x2, x);
+		}
+		/* Check clipping. */
+		x1 = Clamp(x1, 0, dpi->width);
+		x2 = Clamp(x2, x1, dpi->width);
+		if (x2 < 0 || x1 == x2) {
+			++y;
+			continue;
+		}
+		/* Fill line y from x1 to x2. */
```

Comment:
The entire filling region thing is wrong, it does need to collect all intersection points and fill between pairs, otherwise many concave shapes can't render correctly.